subprojects/docs/src/docs/userguide/feature_variants.adoc

```diff
@@ -0,0 +1,226 @@
+It will also, by convention, create a `Jar` task to bundle the classes built from this feature source set, using a classifier corresponding to the kebab-case name of the feature:
+
+.Declaring a feature variant using a separate source set
+====
+include::sample[dir="java-feature-variant/producer-separate-sourceset/groovy",files="build.gradle[tags=register_variant]"]
+include::sample[dir="java-feature-variant/producer-separate-sourceset/kotlin",files="build.gradle.kts[tags=register_variant]"]
+====
+
+[[sec::publishing_feature_variants]]
+== Publishing feature variants
+
+[WARNING]
+====
+Depending on the metadata file format, publishing feature variants may be lossy:
+
+- using {metadata-file-spec}[Gradle metadata], everything is published and consumers will get the full benefit of feature variants
+- using POM metadata (Maven), feature variants are published as **optional dependencies** and artifacts of feature variants are published with different _classifiers_
+- using Ivy metadata, feature variants are lost
+====
+
+Publishing feature variants is supported using the `maven-publish` plugin only.
+The Java Plugin (or Java Library Plugin) will take care of registering the additional variants for you, so there's no additional configuration required, only the regular publications:
+
+.Publishing a component with feature variants
+====
+include::sample[dir="java-feature-variant/producer/groovy",files="build.gradle[tags=plugins,publishing]"]
+include::sample[dir="java-feature-variant/producer/kotlin",files="build.gradle.kts[tags=plugins,publishing]"]
+====
+
+[[sec::consuming_feature_variants]]
+== Dependencies on feature variants
+
+[WARNING]
+====
+As mentioned earlier, feature variants can be lossy when published.
+As a consequence, a consumer can depend on a feature variant only in two cases:
+- with a project dependency (in a multi-project build)
+- with Gradle metadata enabled (the publisher MUST have published using Gradle metadata)
+====
+
+A consumer can specify that it needs a specific feature of a producer by declaring required capabilities.
+For example, if a producer declares a "MySQL support" feature like this:
+
+.A library declaring a feature to support MySQL
+====
+include::sample[dir="java-feature-variant/requiring-features/groovy/producer",files="build.gradle[tags=producer]"]
+include::sample[dir="java-feature-variant/requiring-features/kotlin/producer",files="build.gradle.kts[tags=producer]"]
+====
+
+Then the consumer can declare a dependency on the MySQL support feature by doing this:
+
+.Consuming specific features in a multi-project build
+====
+include::sample[dir="java-feature-variant/requiring-features/groovy/consumer",files="build.gradle[tags=consumer]"]
+include::sample[dir="java-feature-variant/requiring-features/kotlin/consumer",files="build.gradle.kts[tags=consumer]"]
+====
+
```

Comment:
I would explicitly state that this second dependency will cause `mysql-connector-java` to be brought into the `runtimeClasspath` transitively.

Reply:
added